File: python_project/frontend/pages/ocr_invoice.py
import streamlit as st
import requests
from fpdf import FPDF
from datetime import datetime
import tempfile

API_OCR = "http://localhost:8000/ocr/extract"
API_SELL = "http://localhost:8000/inventory/sell"
API_SEARCH = "http://localhost:8000/search"
API_INVENTORY = "http://localhost:8000/inventory/all"

def render_ocr_invoice_page():
    st.title("💊 OCR + Invoice Generator")

    mode = st.radio("Choose Mode", ["Manual Entry", "OCR via Prescription"])

    meds = []
    patient = doctor = clinic = ""
    date = datetime.today()

    if mode == "Manual Entry":
        st.subheader("📝 Patient and Prescription Details")
        patient = st.text_input("Patient Name")
        doctor = st.text_input("Doctor Name")
        clinic = st.text_input("Clinic Name")
        date = st.date_input("Date", value=datetime.today())

        st.subheader("💊 Add Medicines")
        num_meds = st.number_input("Number of Medicines", min_value=1, max_value=10, step=1)
        for i in range(num_meds):
            name = st.text_input(f"Medicine {i+1} Name", key=f"name_{i}")
            quantity = st.number_input(f"Quantity for {name or f'Medicine {i+1}'}", min_value=1, key=f"qty_{i}")
            if name:
                meds.append({"name": name, "quantity": quantity})

    elif mode == "OCR via Prescription":
        st.subheader("📤 Upload Prescription Image")
        image = st.file_uploader("Upload Image", type=["png", "jpg", "jpeg"])

        if image:
            with st.spinner("Extracting text..."):
                res = requests.post(API_OCR, files={"file": image})
                if res.status_code == 200:
                    data = res.json()
                    patient = st.text_input("Patient Name", value=data.get("Patient's Name", ""))
                    doctor = st.text_input("Doctor Name", value=data.get("Doctor's Name", ""))
                    clinic = st.text_input("Clinic Name", value=data.get("Clinic Name", ""))
                    date = st.date_input("Date", value=datetime.today())

                    extracted = data.get("Medicines Prescribed", [])
                    st.markdown("### 📋 Extracted Medicines")
                    for i, extracted_name in enumerate(extracted):
                        name = st.text_input(f"Medicine {i+1} Name", value=extracted_name, key=f"ocr_name_{i}")
                        quantity = st.number_input(f"Quantity for {name}", min_value=1, key=f"ocr_qty_{i}")
                        if name:
                            meds.append({"name": name, "quantity": quantity})

        st.markdown("### ➕ Add Extra Medicines (Optional)")
        num_extra = st.number_input("Extra medicines to add", min_value=0, max_value=5, step=1, key="extra_ocr_count")
        for j in range(num_extra):
            extra_name = st.text_input(f"Extra Medicine {j+1} Name", key=f"extra_ocr_name_{j}")
            extra_qty = st.number_input(f"Quantity for {extra_name or f'Med {j+1}'}", min_value=1, key=f"extra_ocr_qty_{j}")
            if extra_name:
                meds.append({"name": extra_name, "quantity": extra_qty})

    st.divider()
    st.subheader("🔍 Check Availability and Alternatives")

    final_meds = []
    if meds:
        try:
            inv = requests.get(API_INVENTORY).json()
            inventory_names = {m["name"].lower(): m for m in inv}
        except:
            st.error("❌ Could not connect to inventory API.")
            return

        for med in meds:
            name = med["name"].strip()
            qty = med["quantity"]
            found = inventory_names.get(name.lower())

            if found and found["quantity"] >= qty:
                st.success(f"✅ {name} is available (Qty: {found['quantity']})")
                final_meds.append({"name": name, "quantity": qty})
            else:
                if found:
                    st.warning(f"⚠️ {name} in stock: {found['quantity']} < requested {qty}")
                else:
                    st.warning(f"❌ {name} not in inventory")

                alt_res = requests.post(f"{API_SEARCH}/similar", json={"medicine_name": name, "top_k": 10})
                if alt_res.status_code == 200:
                    try:
                        alts = alt_res.json()
                        options = []
                        used_names = set()

                        for alt in alts:
                            alt_name = alt["name"]
                            if alt_name.lower() in used_names:
                                continue
                            used_names.add(alt_name.lower())

                            match = inventory_names.get(alt_name.lower())
                            if match and match["quantity"] > 0:
                                options.append((alt_name, match["quantity"]))
                            if len(options) >= 5:
                                break

                        if options:
                            st.success("✅ Found top vector alternatives:")
                            for alt_name, qty_available in options:
                                st.markdown(f"- {alt_name} (In stock: {qty_available})")

                            alt_only = [opt[0] for opt in options]
                            selected_alt = st.selectbox(f"Select alternative for {name}", alt_only, key=f"alt_select_{name}")
                            max_qty = dict(options)[selected_alt]
                            alt_qty = st.number_input(f"Quantity for {selected_alt}", min_value=1, max_value=max_qty, key=f"alt_qty_{name}")
                            final_meds.append({"name": selected_alt, "quantity": alt_qty})
                        else:
                            st.error(f"⚠️ No in-stock vector alternatives found for {name}")
                    except Exception as e:
                        st.error(f"❌ Error parsing alternatives: {e}")
                else:
                    st.error(f"❌ Vector search failed: {alt_res.status_code}")
                    st.text(alt_res.text)

    if final_meds and st.button("🧾 Generate Invoice and Update Stock"):
        payload = {"medicines": final_meds}
        with st.spinner("Processing..."):
            res = requests.post(API_SELL, json=payload)

        if res.status_code == 200:
            invoice_data = res.json()["invoice"]
            st.success("✅ Inventory updated and invoice ready!")

            pdf = FPDF()
            pdf.add_page()
            pdf.set_font("Arial", size=12)
            pdf.cell(200, 10, txt="Pharmacy Invoice", ln=1, align="C")
            pdf.ln(10)
            pdf.cell(200, 10, txt=f"Patient: {patient}", ln=1)
            pdf.cell(200, 10, txt=f"Doctor: {doctor}", ln=1)
            pdf.cell(200, 10, txt=f"Clinic: {clinic}", ln=1)
            pdf.cell(200, 10, txt=f"Date: {date}", ln=1)
            pdf.ln(10)

            pdf.set_font("Arial", "B", size=12)
            pdf.cell(70, 10, "Medicine", border=1)
            pdf.cell(30, 10, "Qty", border=1)
            pdf.cell(40, 10, "Unit Price", border=1)
            pdf.cell(40, 10, "Subtotal", border=1)
            pdf.ln()

            total = 0
            pdf.set_font("Arial", size=12)
            for item in invoice_data["items"]:
                pdf.cell(70, 10, item["name"], border=1)
                pdf.cell(30, 10, str(item["quantity"]), border=1)
                pdf.cell(40, 10, f"Rs. {item['unit_price']:.2f}", border=1)
                pdf.cell(40, 10, f"Rs. {item['subtotal']:.2f}", border=1)
                pdf.ln()
                total += item["subtotal"]

            pdf.set_font("Arial", "B", size=12)
            pdf.cell(140, 10, "Total", border=1)
            pdf.cell(40, 10, f"Rs. {total:.2f}", border=1)
            pdf.ln()

            tmp = tempfile.NamedTemporaryFile(delete=False, suffix=".pdf")
            pdf.output(tmp.name)
            with open(tmp.name, "rb") as f:
                st.download_button("📥 Download Invoice PDF", f, file_name="invoice.pdf", mime="application/pdf")
        else:
            st.error("❌ Failed to generate invoice or update stock.")

File: Mern_project/frontend/src/pages/OCRInvoice.js
import React, { useState } from 'react';
import {
  Card,
  Radio,
  Input,
  Button,
  Upload,
  DatePicker,
  InputNumber,
  Select,
  Space,
  Alert,
  message,
  Spin,
  Divider,
  Typography,
  Row,
  Col
} from 'antd';
import {
  UploadOutlined,
  FileTextOutlined,
  DownloadOutlined,
  CheckCircleOutlined,
  ExclamationCircleOutlined
} from '@ant-design/icons';
import { ocrAPI, inventoryAPI, searchAPI } from '../services/api';
import jsPDF from 'jspdf';

const { Title, Text } = Typography;
const { TextArea } = Input;

const OCRInvoice = () => {
  const [mode, setMode] = useState('manual');
  const [loading, setLoading] = useState(false);
  const [extractedData, setExtractedData] = useState(null);
  const [medicines, setMedicines] = useState([]);
  const [selectedMedicines, setSelectedMedicines] = useState([]);
  const [inventory, setInventory] = useState([]);
  const [formData, setFormData] = useState({
    patientName: '',
    doctorName: '',
    clinicName: '',
    date: new Date(),
    medicines: []
  });

  // Handle file upload for OCR
  const handleFileUpload = async (file) => {
    try {
      setLoading(true);
      const response = await ocrAPI.extractTextFromImage(file);
      setExtractedData(response.data);
      
      // Pre-fill form with extracted data
      setFormData(prev => ({
        ...prev,
        patientName: response.data["Patient's Name"] || '',
        doctorName: response.data["Doctor's Name"] || '',
        clinicName: response.data["Clinic Name"] || '',
        medicines: response.data["Medicines Prescribed"] || []
      }));
      
      message.success('Text extracted successfully!');
    } catch (error) {
      message.error('Failed to extract text from image');
      console.error('OCR Error:', error);
    } finally {
      setLoading(false);
    }
  };

  // Fetch inventory data
  React.useEffect(() => {
    const fetchInventory = async () => {
      try {
        const response = await inventoryAPI.getAllMedicines();
        setInventory(response.data);
      } catch (error) {
        console.error('Error fetching inventory:', error);
      }
    };
    fetchInventory();
  }, []);

  // Add medicine to prescription
  const addMedicine = () => {
    setMedicines([...medicines, { name: '', quantity: 1 }]);
  };

  // Update medicine in prescription
  const updateMedicine = (index, field, value) => {
    const updated = [...medicines];
    updated[index] = { ...updated[index], [field]: value };
    setMedicines(updated);
  };

  // Remove medicine from prescription
  const removeMedicine = (index) => {
    setMedicines(medicines.filter((_, i) => i !== index));
  };

  // Check medicine availability and find alternatives
  const checkAvailability = async (medicineName, quantity) => {
    const found = inventory.find(med => 
      med.name.toLowerCase() === medicineName.toLowerCase()
    );

    if (found && found.quantity >= quantity) {
      return { available: true, medicine: found };
    }

    // Search for alternatives
    try {
      const response = await searchAPI.searchSimilarMedicines(medicineName, 5);
      const alternatives = response.data
        .map(alt => inventory.find(med => med.name === alt.name))
        .filter(Boolean)
        .filter(med => med.quantity > 0);
      
      return { 
        available: false, 
        alternatives: alternatives.slice(0, 3) 
      };
    } catch (error) {
      console.error('Error finding alternatives:', error);
      return { available: false, alternatives: [] };
    }
  };

  // Process prescription and generate invoice
  const processPrescription = async () => {
    try {
      setLoading(true);
      
      // Check availability for all medicines
      const availabilityResults = await Promise.all(
        medicines.map(async (med) => {
          const result = await checkAvailability(med.name, med.quantity);
          return { ...med, ...result };
        })
      );

      // Filter available medicines
      const availableMedicines = availabilityResults
        .filter(result => result.available)
        .map(result => ({
          name: result.name,
          quantity: result.quantity
        }));

      if (availableMedicines.length === 0) {
        message.error('No medicines are available in stock');
        return;
      }

      // Process sale
      const response = await inventoryAPI.sellMedicines(availableMedicines);
      const invoice = response.data.invoice;

      // Generate PDF
      generateInvoicePDF(invoice);
      
      message.success('Invoice generated and stock updated successfully!');
      setMedicines([]);
      
    } catch (error) {
      message.error('Failed to process prescription');
      console.error('Processing Error:', error);
    } finally {
      setLoading(false);
    }
  };

  // Generate PDF invoice
  const generateInvoicePDF = (invoiceData) => {
    const pdf = new jsPDF();
    
    // Header
    pdf.setFontSize(20);
    pdf.text('Pharmacy Invoice', 105, 20, { align: 'center' });
    
    // Patient and prescription details
    pdf.setFontSize(12);
    pdf.text(`Patient: ${formData.patientName}`, 20, 40);
    pdf.text(`Doctor: ${formData.doctorName}`, 20, 50);
    pdf.text(`Clinic: ${formData.clinicName}`, 20, 60);
    pdf.text(`Date: ${formData.date.toLocaleDateString()}`, 20, 70);
    
    // Items table header
    pdf.setFontSize(12);
    pdf.setFont(undefined, 'bold');
    pdf.text('Medicine', 20, 90);
    pdf.text('Qty', 100, 90);
    pdf.text('Unit Price', 130, 90);
    pdf.text('Subtotal', 170, 90);
    
    // Items
    pdf.setFont(undefined, 'normal');
    let y = 100;
    invoiceData.items.forEach(item => {
      pdf.text(item.name, 20, y);
      pdf.text(item.quantity.toString(), 100, y);
      pdf.text(`₹${item.unit_price.toFixed(2)}`, 130, y);
      pdf.text(`₹${item.subtotal.toFixed(2)}`, 170, y);
      y += 10;
    });
    
    // Total
    pdf.setFont(undefined, 'bold');
    pdf.text('Total', 130, y + 10);
    pdf.text(`₹${invoiceData.total.toFixed(2)}`, 170, y + 10);
    
    // Download
    pdf.save('invoice.pdf');
  };

  const uploadProps = {
    name: 'file',
    accept: 'image/*',
    beforeUpload: handleFileUpload,
    showUploadList: false,
  };

  return (
    <div>
      <Title level={2}>💊 OCR + Invoice Generator</Title>

      <Card style={{ marginBottom: '24px' }}>
        <Radio.Group 
          value={mode} 
          onChange={(e) => setMode(e.target.value)}
          style={{ marginBottom: '16px' }}
        >
          <Radio.Button value="manual">Manual Entry</Radio.Button>
          <Radio.Button value="ocr">OCR via Prescription</Radio.Button>
        </Radio.Group>

        {mode === 'manual' ? (
          <div>
            <Title level={4}>📝 Patient and Prescription Details</Title>
            <Row gutter={16}>
              <Col span={8}>
                <Input
                  placeholder="Patient Name"
                  value={formData.patientName}
                  onChange={(e) => setFormData({...formData, patientName: e.target.value})}
                />
              </Col>
              <Col span={8}>
                <Input
                  placeholder="Doctor Name"
                  value={formData.doctorName}
                  onChange={(e) => setFormData({...formData, doctorName: e.target.value})}
                />
              </Col>
              <Col span={8}>
                <Input
                  placeholder="Clinic Name"
                  value={formData.clinicName}
                  onChange={(e) => setFormData({...formData, clinicName: e.target.value})}
                />
              </Col>
            </Row>
          </div>
        ) : (
          <div>
            <Title level={4}>📤 Upload Prescription Image</Title>
            <Upload {...uploadProps}>
              <Button icon={<UploadOutlined />} loading={loading}>
                Upload Image
              </Button>
            </Upload>
            
            {extractedData && (
              <div style={{ marginTop: '16px' }}>
                <Alert
                  message="Extracted Data"
                  description={
                    <div>
                      <p><strong>Patient:</strong> {extractedData["Patient's Name"] || 'Not found'}</p>
                      <p><strong>Doctor:</strong> {extractedData["Doctor's Name"] || 'Not found'}</p>
                      <p><strong>Clinic:</strong> {extractedData["Clinic Name"] || 'Not found'}</p>
                      <p><strong>Medicines:</strong> {extractedData["Medicines Prescribed"]?.join(', ') || 'None found'}</p>
                    </div>
                  }
                  type="success"
                  showIcon
                />
              </div>
            )}
          </div>
        )}
      </Card>

      <Card title="💊 Add Medicines" style={{ marginBottom: '24px' }}>
        <Button 
          type="dashed" 
          onClick={addMedicine}
          style={{ marginBottom: '16px' }}
        >
          + Add Medicine
        </Button>

        {medicines.map((medicine, index) => (
          <Card 
            key={index} 
            size="small" 
            style={{ marginBottom: '8px' }}
            actions={[
              <Button 
                type="link" 
                danger 
                onClick={() => removeMedicine(index)}
              >
                Remove
              </Button>
            ]}
          >
            <Row gutter={16} align="middle">
              <Col span={12}>
                <Input
                  placeholder="Medicine Name"
                  value={medicine.name}
                  onChange={(e) => updateMedicine(index, 'name', e.target.value)}
                />
              </Col>
              <Col span={8}>
                <InputNumber
                  min={1}
                  placeholder="Quantity"
                  value={medicine.quantity}
                  onChange={(value) => updateMedicine(index, 'quantity', value)}
                  style={{ width: '100%' }}
                />
              </Col>
              <Col span={4}>
                <Button 
                  type="primary" 
                  onClick={async () => {
                    const result = await checkAvailability(medicine.name, medicine.quantity);
                    if (result.available) {
                      message.success(`${medicine.name} is available`);
                    } else {
                      message.warning(`${medicine.name} not available`);
                    }
                  }}
                >
                  Check
                </Button>
              </Col>
            </Row>
          </Card>
        ))}
      </Card>

      <Card title="🔍 Check Availability and Alternatives">
        <Space direction="vertical" style={{ width: '100%' }}>
          {medicines.map((medicine, index) => (
            <div key={index}>
              <Text strong>{medicine.name}</Text>
              <Text type="secondary"> (Qty: {medicine.quantity})</Text>
            </div>
          ))}
        </Space>

        <Divider />

        <Button
          type="primary"
          size="large"
          icon={<FileTextOutlined />}
          onClick={processPrescription}
          loading={loading}
          disabled={medicines.length === 0}
        >
          Generate Invoice and Update Stock
        </Button>
      </Card>
    </div>
  );
};

export default OCRInvoice;
